src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java

```diff
@@ -101,63 +117,104 @@ public void initialise() {
      *
      * @return option of the liquid block found in reach, empty if none was found
      */
-    private Optional<Vector3i> getLiquidInReach(final Vector3f start, final Vector3f direction, int distance) {
-        Vector3f location = new Vector3f(start);
-        Vector3f normalizedDirection = new Vector3f(direction).normalize();
-        for (int i = 0; i < distance; i++) {
-            location.add(normalizedDirection);
-            Vector3i intLocation = new Vector3i(location, RoundingMode.HALF_DOWN);
-            Block block = worldProvider.getBlock(intLocation);
-            if (block.isLiquid()) {
-                return Optional.of(intLocation);
-            }
-            if (!block.isPenetrable()) {
-                return Optional.empty();
-            }
+    private Optional<Vector3i> getLiquidInReach(final Vector3f start, final Vector3f direction, EntityRef character, float distance) {
+        HitResult hitResult = physics.rayTrace(start, direction, distance, Sets.newHashSet(character), PHYSICSFILTER);
```

Comment:
I don't understand, why we're checking for a raytrace hit here :thinking: can you elaborate a bit?

Reply:
When the player clicks with an item, the LocalPlayer casts a ray in the direction the player is facing to determine what this action hits (e.g. where to place it if it's a right click with a block item), but that ray doesn't collide with liquids. In this case, it should be possible to place liquids on other liquids and must be possible to target liquids for removal, therefore the FluidAuthoritySystem does its own raytracing that includes liquids to determine what block to affect.

Reply:
Interesting, I thought I saw it done differently in other places :thinking: Something about viewing target of the player or something, don't recall exactly unfortunately... Maybe the raytracing was more hidden in those cases through some helper functions from the engine or iota modules :thinking:

Reply:
There is another event fired by the engine that includes the information about hit location, that's just not usable here because of the issue with it not including liquids.

Reply:
Ohh okay, understood now :+1: